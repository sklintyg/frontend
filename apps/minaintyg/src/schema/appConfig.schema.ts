import { z } from 'zod'

export const appConfigLinkSchema = z.object({
  key: z.string(),
  url: z.string(),
  text: z.string(),
  target: z.string(),
})

export const appConfigSchema = z.object({
  useMinifiedJavascript: z.boolean(),
  version: z.string(),
  buildNumber: z.string(),
  elva77MainUrl: z.string(),
  elva77LoginUrl: z.string(),
  applicationLogoutUrl: z.string(),
  miUserSurveyUrl: z.string(),
  miUserSurveyVersion: z.string(),
  miUserSurveyDateFrom: z.string(),
  miUserSurveyDateTo: z.string(),
  knownRecipients: z.array(z.object({ id: z.string(), name: z.string(), trusted: z.boolean() })),
  links: z.record(appConfigLinkSchema),
  banners: z.array(z.unknown()),
})

export type AppConfig = z.infer<typeof appConfigSchema>
export type AppConfigLink = z.infer<typeof appConfigLinkSchema>
