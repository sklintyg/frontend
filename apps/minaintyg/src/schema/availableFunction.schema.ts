import { z } from 'zod'

export const AvailableFunctionType = z.enum(['SEND_CERTIFICATE'])
export const AvailableFunctionInformationType = z.enum([''])

export const availableFunctionInformationSchema = z.object({
  type: AvailableFunctionInformationType,
  id: z.string(),
})

export const availableFunctionSchema = z.object({
  body: z.string().nullable(),
  description: z.string().nullable(),
  name: z.string().nullable(),
  title: z.string().nullable(),
  type: AvailableFunctionType,
  information: z.array(availableFunctionInformationSchema),
})

export type AvailableFunction = z.infer<typeof availableFunctionSchema>
