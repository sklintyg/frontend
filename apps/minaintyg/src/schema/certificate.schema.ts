import { z } from 'zod'

export const CertificateStatusEnum = z.enum(['NEW', 'REPLACED', 'SENT', 'NOT_SENT'])

export const certificateEventSchema = z.object({
  timestamp: z.string().datetime(),
  certificateId: z.string(),
  description: z.string(),
})

export const certificateSummarySchema = z.object({ label: z.string(), value: z.string() }).optional()

export const certificateTypeSchema = z.object({
  id: z.string(),
  version: z.string(),
  description: z.string().optional(),
})

export const certificateIssuerSchema = z.object({
  name: z.string(),
  phoneNumber: z.string(),
})

export const certificateUnitSchema = z.object({
  id: z.string(),
  name: z.string(),
  address: z.string(),
})

export const certificateMetadataSchema = z.object({
  issuer: certificateIssuerSchema,
  unit: certificateUnitSchema,
  name: z.string(),
  events: z.array(certificateEventSchema),
  statuses: z.array(CertificateStatusEnum),
  id: z.string(),
  issued: z.string().datetime(),
  type: certificateTypeSchema,
  summary: certificateSummarySchema,
})

export const certificateContentSchema = z.object({ heading: z.string(), body: z.string() })

export const certificateSchema = z.object({
  metadata: certificateMetadataSchema,
  content: z.array(certificateContentSchema),
})

export type CertificateContent = z.infer<typeof certificateContentSchema>
export type CertificateEvent = z.infer<typeof certificateEventSchema>
export type CertificateIssuer = z.infer<typeof certificateIssuerSchema>
export type CertificateMetadata = z.infer<typeof certificateMetadataSchema>
export type CertificateStatus = z.infer<typeof CertificateStatusEnum>
export type CertificateSummary = z.infer<typeof certificateSummarySchema>
export type CertificateUnit = z.infer<typeof certificateUnitSchema>
export type Certificate = z.infer<typeof certificateSchema>
