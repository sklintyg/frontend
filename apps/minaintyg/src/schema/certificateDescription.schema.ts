import { z } from 'zod'

export const certificateDescriptionSchema = z.object({
  id: z.string(),
  label: z.string(),
  description: z.string(),
  detailedDescription: z.string(),
  issuerTypeId: z.string(),
  cssPath: z.string(),
  scriptPath: z.string(),
  dependencyDefinitionPath: z.string(),
  defaultRecipient: z.string(),
  deprecated: z.boolean(),
  displayDeprecated: z.boolean(),
})

export type CertificateDescription = z.infer<typeof certificateDescriptionSchema>
