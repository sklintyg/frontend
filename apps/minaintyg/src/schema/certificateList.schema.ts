import { z } from 'zod'

export enum CertificateStatus {
  NEW = 'Nytt',
  REPLACED = 'Ers√§tter intyg',
  SENT = 'Skickat',
  NOT_SENT = 'Ej skickat',
}

export const certificateListSummarySchema = z.object({ label: z.string(), value: z.string() }).optional()

export const certificateListEventSchema = z.object({
  timestamp: z.string().datetime(),
  certificateId: z.string(),
  description: z.string(),
})

export const certificateListIssuerSchema = z.object({ name: z.string() })
export const certificateUnitSchema = z.object({ id: z.string(), name: z.string() })
export const certificateTypeSchema = z.object({
  id: z.string(),
  name: z.string(),
  version: z.string(),
})

export const certificateListItemSchema = z.object({
  summary: certificateListSummarySchema,
  issuer: certificateListIssuerSchema,
  unit: certificateUnitSchema,
  events: z.array(certificateListEventSchema),
  statuses: z.array(z.nativeEnum(CertificateStatus)),
  id: z.string(),
  issued: z.string().datetime(),
  type: certificateTypeSchema,
})

export type CertificateListEvent = z.infer<typeof certificateListEventSchema>
export type CertificateListIssuer = z.infer<typeof certificateListIssuerSchema>
export type CertificateListUnit = z.infer<typeof certificateUnitSchema>
export type CertificateListSummary = z.infer<typeof certificateListSummarySchema>
export type CertificateListItem = z.infer<typeof certificateListItemSchema>
