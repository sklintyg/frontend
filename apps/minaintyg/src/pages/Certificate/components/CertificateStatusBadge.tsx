import { IDSBadge } from 'ids-react-ts'
import type { ComponentProps } from 'react'
import type { CertificateStatus } from '../../../schema/certificate.schema'
import { CertificateStatusEnum } from '../../../schema/certificate.schema'
import { getStatusBadgeLabel } from '../utils/getStatusBadgeLabel'

export function getBadgeType(status: CertificateStatus): ComponentProps<typeof IDSBadge>['type'] {
  switch (status) {
    case CertificateStatusEnum.enum.NEW:
      return 'primary'
    case CertificateStatusEnum.enum.SENT:
      return 'success'
    case CertificateStatusEnum.enum.NOT_SENT:
      return 'error'
    case CertificateStatusEnum.enum.REPLACED:
      return 'neutral'
    default:
      return 'secondary'
  }
}

export function CertificateStatusBadge({ status }: { status: CertificateStatus }) {
  const label = getStatusBadgeLabel(status)

  if (!label) {
    return null
  }

  return (
    <IDSBadge className="whitespace-nowrap" type={getBadgeType(status)}>
      {label}
    </IDSBadge>
  )
}
