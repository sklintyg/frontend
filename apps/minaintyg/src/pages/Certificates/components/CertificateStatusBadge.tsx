import { IDSBadge } from '@frontend/ids-react-ts'
import { ComponentProps } from 'react'
import { CertificateStatus } from '../../../schema/certificateList.schema'

export const getBadgeType = (status: CertificateStatus): ComponentProps<typeof IDSBadge>['type'] => {
  switch (status) {
    case CertificateStatus.NEW:
      return 'primary'
    case CertificateStatus.SENT:
      return 'success'
    case CertificateStatus.NOT_SENT:
      return 'error'
    default:
      return 'secondary'
  }
}

export function CertificateStatusBadge({ status }: { status: CertificateStatus }) {
  return (
    <IDSBadge className="[&:not(:last-child)]:mb-2.5" type={getBadgeType(status)}>
      {status}
    </IDSBadge>
  )
}
