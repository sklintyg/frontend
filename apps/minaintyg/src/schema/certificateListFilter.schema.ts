import { z } from 'zod'
import { CertificateStatus } from './certificateList.schema'

export const certificateFilterOptionsSchema = z.object({
  total: z.number(),
  statuses: z.array(z.nativeEnum(CertificateStatus)),
  units: z.array(z.string()),
  certificateTypes: z.array(z.string()),
  years: z.array(z.string()),
})

type SelectedOptions<T> = { [X in keyof T]: T[X] extends Array<unknown> ? T[X][number] : T[X] }

export type CertificateFilterOptions = z.infer<typeof certificateFilterOptionsSchema>
export type CertificateSelectedOptions = SelectedOptions<CertificateFilterOptions>
