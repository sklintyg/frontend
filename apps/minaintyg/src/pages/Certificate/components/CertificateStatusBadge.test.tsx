import { render, screen } from '@testing-library/react'
import { CertificateStatus, CertificateStatusEnum } from '../../../schema/certificate.schema'
import { CertificateStatusBadge, getBadgeType } from './CertificateStatusBadge'

it('Should render component', () => {
  const { baseElement } = render(<CertificateStatusBadge status={CertificateStatusEnum.enum.NEW} />)
  expect(baseElement).toMatchSnapshot()
})

it.each([
  [CertificateStatusEnum.enum.NEW, 'primary'],
  [CertificateStatusEnum.enum.SENT, 'success'],
  [CertificateStatusEnum.enum.NOT_SENT, 'error'],
  [CertificateStatusEnum.enum.REPLACED, 'secondary'],
])('Should return %s for certificate status %s', (status, expected) => {
  expect(getBadgeType(status)).toBe(expected)
})

it.each([
  [CertificateStatusEnum.enum.NEW, 'Nytt'],
  [CertificateStatusEnum.enum.REPLACED, 'ErsÃ¤tter intyg'],
  [CertificateStatusEnum.enum.SENT, 'Skickat'],
  [CertificateStatusEnum.enum.NOT_SENT, 'Ej skickat'],
])('Should print %s for status %s', (status, expected) => {
  render(<CertificateStatusBadge status={status} />)
  expect(screen.getByText(expected)).toBeInTheDocument()
})

it('Should not return a badge for unknown status', () => {
  const { container } = render(<CertificateStatusBadge status={'FOASDASD' as CertificateStatus} />)
  expect(container).toBeEmptyDOMElement()
})
