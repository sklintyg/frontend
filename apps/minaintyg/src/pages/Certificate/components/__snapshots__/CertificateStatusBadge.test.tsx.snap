// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render component 1`] = `
<div>
  <div
    class="ids-badge ids-badge--primary whitespace-nowrap"
  >
    <div
      class="ids-badge__inner"
    >
      Nytt
    </div>
  </div>
</div>
`;
